substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -401,29 +446,53 @@ pub mod pallet {
             <T as Config>::RestrictedOrigin::ensure_origin(origin)?;
             Self::_approve_solution_provider(solution_provider_id, approve)
         }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        pub fn bill_contract_for_block(
+            origin: OriginFor<T>,
+            contract_id: u64,
+            block_number: T::BlockNumber,
+        ) -> DispatchResultWithPostInfo {
+            let _account_id = ensure_signed(origin)?;
+            Self::_bill_contract_for_block(contract_id, block_number)
+        }
     }
 
     #[pallet::hooks]
     impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
-        fn on_finalize(block: T::BlockNumber) {
-            match Self::_bill_contracts_at_block(block) {
-                Ok(_) => {
-                    log::info!(
-                        "types::NodeContract billed successfully at block: {:?}",
-                        block
-                    );
-                }
-                Err(err) => {
-                    log::info!(
-                        "types::NodeContract billed failed at block: {:?} with err {:?}",
-                        block,
-                        err
-                    );
-                }
+        fn offchain_worker(block_number: T::BlockNumber) {
+            let current_block_u64: u64 = block_number.saturated_into::<u64>();
+            let mut contracts = Self::get_failed_contract_ids_from_storage();
+            if !contracts.is_empty() {
+                log::info!(
```

Comment:
Bill is never executed for the failed ones?

Reply:
Yes it is. get_failed_contract_ids_from_storage fetches the failed contracts from the block before and retries them here.